Fix up sporadically failing tests and ensure tests can run independently

zenspider opened the pull request:
Most of the info is in the commit messages... I should whip up some way to collate them into something I can paste here. 😕

zenspider committed:
Fixes loading of Telemetry::DataSeries (and neighbors) in isolation.

Fixes telemetry tests when run individually.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Use Inspec::VERSION directly to decouple from bundler.

Can also try to use the bundler version first and then fall back to
VERSION... but VERSION is absolutely free.

Happy to push an edit to this to remove the comment or fold it in.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Functional helper let changes.

Moved exec_inspec to inspec_path.
Added new exec_inspec that invokes ruby w/ -Ilib (expanded).

Decouples from bundler and/or needing inspec-bin to be installed.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Minor cleanup and a todo test.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Changed cmp matcher to use send.

obj.method(name).call(*args) == obj.send(name, *args)

```
Calculating -------------------------------------
         method+call      2.561M (± 2.2%) i/s -     12.816M in   5.006835s
                send      6.299M (± 2.4%) i/s -     31.501M in   5.004045s

Comparison:
                send:  6299002.5 i/s
         method+call:  2560909.2 i/s - 2.46x  slower
```

There aren't any direct tests for this? But I forcefully failed 2
functionals and then verified they passed after my change.

Not sure how (or if) to write tests against rspec matchers in a
meaningful way, even tho all I really want to test is a plain ole
method in it. I'll poke to see if it is possible / practical.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Tests for cmp.

Doesn't test any of the aux methods.

This uncovers ~3 errors in the code. Not fixed.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed 3 failures in cmp matcher.

+ float? comparison can raise a TypeError
+ octal? comparison was allowing non-octal values (which cast to 0)
+ symbol comparison was casting to a string, but then doing an == check instead of casecmp.

The latter seems optional, but consistent with the intent of cmp.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed missing requires

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed rake test:resources task. The path was no longer valid.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed rake test:isolated task. The glob was far too limited.

Extracted a GLOBS const so it can be reused and only updated in one
place.

Added test:list that prints out all known test files.

Changed test:isolated to use the glob, AND to use system (sh would
fail out on the first failing test, so much of the looping mechanism
was redundant. I also changed the loop to run all of the tests
regardless of outcome, so you can see all the failures.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Use RbConfig::CONFIG instead of Gem.ruby_version for plugin_gem_path

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Add minitest/hell and parallelize_me! to the plugin tests.

We already do this in test/helper.rb so we might as well do it here.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fix how we're running inspec to include local lib dir.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed referenced urls so we can find the cited code again.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Turn off rdoc/ri generation in plugins.

As commented, I'm happy to push this to our tests, but it does seem
odd that we'd install rdoc/ri for plugins.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed two test order dependency bugs.

+ Do extra cleanup of loaded_specs for the dependencies.
+ Ensure rake is actually activated.

Fixes #4030.
Fixes #4026.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Added SLOW=1 env var to run slow installer tests, otherwise skipped

+ Added SLOW=1 to appveyor (I hope?) and travis.
+ Off by default.

Speeds up installer tests from ~5.5 seconds to ~0.5 seconds.

Once this is established, we can push this up to the main helper and
generalize if it works out for us.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Punting on 2 tests in inspec-plugin-manager-cli.

These two tests make assumptions that are not valid when bundler is
not being used.

Signed-off-by: Ryan Davis <zenspider@chef.io>

miah commented on .travis.yml:
```diff
@@ -12,6 +12,8 @@ before_install:
 - gem --version
 - bundle --version
+env:
+  - SLOW=1
```
I like this, and that the slow tests get marked as skipped so we don't forget about them =)

zenspider replied:
Exactly. That was the intent.

If you know of a better way to globally set an env on appveyor, I'd appreciate it.

clintoncwolfe commented on lib/inspec/plugin/v2/loader.rb:
```diff
@@ -113,9 +113,8 @@ def plugin_gem_path
     def self.plugin_gem_path
-      # I can't believe there isn't a simpler way of getting this
-      # 2.4.2.p123 => 2.4.0
-      ruby_abi_version = (Gem.ruby_version.segments[0, 2] << 0).join('.')
```
great!

clintoncwolfe commented on lib/matchers/matchers.rb:
```diff
@@ -247,26 +247,26 @@ def try_match(actual, op, expected) # rubocop:disable Metrics/CyclomaticComplexi
     # if actual and expected are strings
     if expected.is_a?(String) && actual.is_a?(String)
       return actual.casecmp(expected) == 0 if op == :==
-      return Gem::Version.new(actual).method(op).call(Gem::Version.new(expected)) if
```
Could you comment as to why these changes are needed to get the tests running more cleanly?

zenspider replied:
They're not. They're part of the general cleanup of stuff that I saw as I went through. The main reason this should go in is this:

```
2896 % ./cmp.rb 
Warming up --------------------------------------
         method+call   154.352k i/100ms
                send   248.925k i/100ms
Calculating -------------------------------------
         method+call      2.603M (± 1.4%) i/s -     13.120M in   5.041969s
                send      6.425M (± 3.1%) i/s -     32.111M in   5.003057s

Comparison:
                send:  6424944.9 i/s
         method+call:  2602626.5 i/s - 2.47x  slower
```

Especially considering that this is on the matcher side of things, and thus, hit by all users.

clintoncwolfe commented on lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb:
```diff
@@ -569,6 +569,8 @@ def test_refuse_install_when_plugin_on_exclusion_list
   end
 
   def test_error_install_with_debug_enabled
+    skip "this test requires bundler to pass" unless defined? ::Bundler
```
Are there test conditions under which Bundler is not available? We require it to run the test suite...

zenspider replied:
We don't require bundler to run the test suite. It runs fine without it with the exception of some of these test assumptions. 

Requiring bundler of our users and especially contributors makes total sense, but the developers on this project shouldn't be beholden to an unnecessary requirement like bundler, especially when it just slows us down.

```
4402 % time ntimes 10 bundle exec ~/.rbenv/versions/2.6.2/bin/ruby -Ilib:test -e 0

real	0m5.506s
user	0m4.504s
sys	0m0.883s
4403 % time ntimes 10 ~/.rbenv/versions/2.6.2/bin/ruby -Ilib:test -e 0

real	0m0.876s
user	0m0.694s
sys	0m0.146s
```

Some of these commits are effectively decoupling our tests from bundler.

clintoncwolfe commented on lib/plugins/inspec-plugin-manager-cli/test/functional/inspec-plugin_test.rb:
```diff
@@ -581,11 +583,13 @@ def test_error_install_with_debug_enabled
   include PluginManagerHelpers
 
   def test_when_a_plugin_can_be_updated
+    skip "this test requires bundler to pass" unless defined? ::Bundler
```
Likewise here, we should have bundler...

clintoncwolfe commented on test/integration/default/controls/audit_spec.rb:
```diff
@@ -14,7 +14,8 @@
 # passed for audit cookbook. If its not, you should do a `bundle install`
-inspec_version = Gem.loaded_specs['inspec'].version.to_s
+inspec_version = Inspec::VERSION
+# or: Gem.loaded_specs['inspec'].version.to_s rescue Inspec::VERSION
```
Could you explain this change?

zenspider replied:
The active code. That is using our VERSION constant instead of finding the same value indirectly all the way through rubygems and back. If the spec isn't loaded, then that will fail out. There's no reason that the spec needs to be loaded for this test to run or pass.

clintoncwolfe commented on test/unit/matchers/matchers_test.rb:
```diff
@@ -0,0 +1,118 @@
+  describe 'cmp matcher' do
+    include RSpec::Matchers
+
+    def rspec_expect value, &block
```
This probably needs in-code comments to explain what is going on. We shouldn't assume maintainers have any particular knowledge of RSpec.

zenspider replied:
fixed

clintoncwolfe commented on test/unit/matchers/matchers_test.rb:
```diff
@@ -0,0 +1,118 @@
+      ::RSpec::Expectations::ExpectationTarget.for(value, block)
+    end
+
+    def assert_cmp exp, act
```
I think `exp` means `expectation`, but I don't know what `act` refers to. Could you use more expressive variable names?

zenspider replied:
"actual". I use this all over (esp in the doco for minitest) so I make assumptions about it's "obviousness". What I really want is a better spoken language that has same-length antonyms (true vs false, local vs remote, etc).

zenspider replied:
fixed

clintoncwolfe commented on test/unit/plugin/v2/installer_test.rb:
```diff
@@ -51,18 +58,19 @@ def teardown
 
     # Clean up any activated gems
     Gem.loaded_specs.delete('inspec-test-fixture')
+    Gem.loaded_specs.delete('ordinal_array')
```
I think I know what this fixes, but just for completeness, could you comment explaining what this is for?

zenspider replied:
What are you looking for here beyond the comment above it? That some tests activate different dependencies and will wind up causing test ordering issues?

clintoncwolfe commented on test/unit/plugin/v2/installer_test.rb:
```diff
@@ -195,13 +209,18 @@ def test_install_a_pinned_gem_from_rubygems_org
   def test_install_a_gem_with_conflicting_depends_from_rubygems_org
     ENV['INSPEC_CONFIG_DIR'] = File.join(@config_dir_path, 'empty')
 
+    spec = Gem::Specification._all.find { |s| s.name == "rake" }
```
Could you explain this?

zenspider replied:
Decouples from bundler.

clintoncwolfe reviewed (commented):
This looks generally good; however on the sporadic failure fixes, there are some that I am not clear on - adding some comments on the PR indicating what fixes what would be very helpful. I also had a clarity concern or two.

Those aren't really blockers; if you add the comments, I can flip to Approve.

zenspider committed:
Omit user plugins for loader test failure.

Fixes #4053

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed assignment to RUBY_VERSION.

Lost in a sea of deprecation warnings.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixes test failure interaction between Loader and Installer tests.

Specifically, if these were run in this order:

+ PluginLoaderTests#test_load_mock_plugin_by_gem
+ PluginInstallerInstallationTests#test_install_a_gem_from_local_file

the latter would fail.

The failure was in test setup. The solution was partially to crib off
of and normalize with LoaderTest setup/teardown and partially to set extra things.

Specifically:

+ HOME wasn't being set so it was finding my ~/.inspec/plugins.json
+ ENV wasn't being reset for Gem.paths
+ Installer wasn't being reset correctly, so plugins were known across tests.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Removed all redundant INSPEC_CONFIG_DIR setup from tests.

Pushed up to setup

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Cleanup long lines for readability / debugging.

Also asked a couple q's in TODO lines.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Skips the one test that hits rspec directly and causes mocha to blow out afterwards

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed another bad test interaction.

This time between:

+ PluginManagerCliDefinitionTests#(?:test_plugin_registered
+ PluginLoaderTests#test_load_mock_plugin_by_gem

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed inconsistent use of MT::S's "expect" method (eg `_`)

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Skip more_permissive_than? file_test because broken only on CI.

Trying to fix file_test failure that only happens on travis.

The "mock" file in question is supposed to have a stat of 644, but
actually has 664 but only on travis-ci.

Signed-off-by: Ryan Davis <zenspider@chef.io>
